src/libraries/Common/src/SourceGenerators/DiagnosticInfo.cs

```diff
@@ -57,4 +89,34 @@ public Diagnostic CreateDiagnostic()
         hashCode = HashHelpers.Combine(hashCode, Location?.GetHashCode() ?? 0);
         return hashCode;
     }
+
+    private static FieldInfo? GetSourceTreeBackingFieldInfo()
+    {
+        FieldInfo? info;
+
+        try
+        {
+            FieldInfo[] fields = typeof(Location).GetFields(BindingFlags.NonPublic);
+            info = typeof(Location).GetField("<SourceTree>k__BackingField", BindingFlags.NonPublic | BindingFlags.Instance);
```

Comment:
I'm not comfortable taking a dependency on this type of implementation detail, particularly in product code. We should instead work with @dotnet/roslyn to come up with a better solution for storing `Location` instances in incremental values that doesn't encapsulate the `Compilation` instance.

cc @stephentoub